src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/scalable/LabeledVariantAnnotationsWalker.java

```diff
@@ -0,0 +1,382 @@
+
+    @Argument(
+            fullName = OMIT_ALLELES_IN_HDF5_LONG_NAME,
+            doc = "If true, omit alleles in output HDF5 files in order to decrease file sizes.",
+            optional = true
+    )
+    boolean omitAllelesInHDF5 = false;
+
+    @Argument(
+            fullName = DO_NOT_GZIP_VCF_OUTPUT_LONG_NAME,
+            doc = "If true, VCF output will not be compressed.",
+            optional = true
+    )
+    boolean doNotGZIPVCFOutput = false;
+
+    private final Set<String> ignoreInputFilterSet = new TreeSet<>();
+    Set<VariantType> variantTypesToExtract;
+    TreeSet<String> resourceLabels = new TreeSet<>();
+
+    File outputAnnotationsFile;
+    VariantContextWriter vcfWriter;
+
+    LabeledVariantAnnotationsData data;
+
+    @Override
+    public void onTraversalStart() {
+
+        ignoreInputFilterSet.addAll(ignoreInputFilters);
+
+        variantTypesToExtract = EnumSet.copyOf(variantTypesToExtractList);
+
+        outputAnnotationsFile = new File(outputPrefix + ANNOTATIONS_HDF5_SUFFIX);
+        final String vcfSuffix = doNotGZIPVCFOutput ? ".vcf" : ".vcf.gz";
+        final File outputVCFFile = new File(outputPrefix + vcfSuffix);
+
+        // TODO this validation method should perhaps be moved outside of the CNV code
+        CopyNumberArgumentValidationUtils.validateOutputFiles(outputAnnotationsFile, outputVCFFile);
+
+        for (final FeatureInput<VariantContext> resource : resources) {
+            final TreeSet<String> trackResourceLabels = resource.getTagAttributes().entrySet().stream()
+                    .filter(e -> e.getValue().equals("true"))
+                    .map(Map.Entry::getKey)
+                    .sorted()
+                    .collect(Collectors.toCollection(TreeSet::new));
+            resourceLabels.addAll(trackResourceLabels);
+            logger.info( String.format("Found %s track: labels = %s", resource.getName(), trackResourceLabels));
+        }
+        resourceLabels.forEach(String::intern);
+
+        if (resourceLabels.contains(LabeledVariantAnnotationsData.SNP_LABEL)) {
+            throw new UserException.BadInput(String.format("The resource label \"%s\" is reserved for labeling variant types.",
+                    LabeledVariantAnnotationsData.SNP_LABEL));
+        }
+
+        data = new LabeledVariantAnnotationsData(annotationNames, resourceLabels, useASAnnotations);
+
+        vcfWriter = createVCFWriter(outputVCFFile);
+        vcfWriter.writeHeader(constructVCFHeader(data.getSortedLabels()));
+
+        afterOnTraversalStart();   // perform additional validation, set modes in child tools, etc.
+    }
+
+    public void afterOnTraversalStart() {
+        // override
+    }
+
+    @Override
+    protected int numberOfPasses() {
+        return 1;
+    }
+
+    @Override
+    public Object onTraversalSuccess() {
+        return null;
+    }
+
+    // TODO maybe clean up all this Triple and metadata business with a class?
```

Comment:
Is there a compelling need to defer this TODO?  It seems worth doing now, IMO.

Reply:
Hmm, after thinking about it more, I removed the TODO and left it as is. This sort of need for a single-use container seems like the perfect time to use an N-tuple.